Description: Minor changes for transition from unixcw to libcw3.
Author: Kamil Ignacak <acerion@wp.pl>
Reviewed-By: Kamal Mostafa <kamal@whence.com>
Bug-Debian: http://bugs.debian.org/652618

--- cwdaemon-0.9.4.orig/cwdaemon.c
+++ cwdaemon-0.9.4/cwdaemon.c
@@ -89,7 +89,7 @@
 #endif
 #include <limits.h>
 
-#include <cwlib.h>
+#include <libcw.h>
 #include "cwdaemon.h"
 
 /* network vars */
@@ -165,6 +165,9 @@ cwdevice cwdevice_lp = {
 cwdevice *cwdev;
 static void playmorsestring (char *x);
 
+static int set_libcw_output (void);
+static void close_libcw (void);
+
 /* catch ^C when running in foreground */
 static RETSIGTYPE
 catchint (int signal)
@@ -299,17 +302,51 @@ tune (int seconds)
 	}
 }
 
-/* initialize unixcw */
+/* (re)set initial parameters of libcw */
 static void
-initmorse (void)
+reset_libcw (void)
 {
+	/* just in case if an old generator exists */
+	close_libcw ();
+
+	set_libcw_output ();
+
 	cw_set_frequency (morse_tone);	
 	cw_set_send_speed (morse_speed);
-	cw_set_soundcard_sound (soundcard_sound);
-	cw_set_console_sound (console_sound);
 	cw_set_volume (morse_volume);
 	cw_set_gap (0);
 }
+ 
+static void
+close_libcw (void)
+{
+	cw_generator_stop ();
+	cw_generator_delete ();
+}
+
+/* set up output of libcw */
+static int
+set_libcw_output (void)
+{
+	int rv = 0;
+	if (soundcard_sound && !console_sound)
+       {
+		rv = cw_generator_new (CW_AUDIO_SOUNDCARD, NULL);
+	}
+	else if (!soundcard_sound && console_sound)
+	{
+		rv = cw_generator_new (CW_AUDIO_CONSOLE, NULL);
+	}
+	else
+	{
+		/* libcw can't do both soundcard and console,
+		   and it has to have one and only one sound
+		   system specified */
+		errmsg ("Sound output specified incorrectly");
+	        rv = CW_FAILURE;
+	}
+	return rv == CW_FAILURE ? -1 : 0;
+}
 
 /* properly parse a 'long' integer */
 static int
@@ -371,7 +408,7 @@ recv_code (void)
 				morse_volume = 70;
 				console_sound = 1;
 				soundcard_sound = 0;
-				initmorse ();
+				reset_libcw ();
 				wordmode = 0;
 				async_abort = 0;
 				cwdev->reset (cwdev);
@@ -559,8 +596,7 @@ recv_code (void)
 				if (valid_sdevice == 1)
 				{
 					debug ("Sound device: %s", message + 2);
-					cw_set_soundcard_sound (soundcard_sound);
-					cw_set_console_sound (console_sound);
+					set_libcw_output ();
 				}
 				break;
 			case 'g':	/* volume */
@@ -858,7 +894,9 @@ main (int argc, char *argv[])
 	}
 	cwdev->desc = keydev;
 
-	initmorse ();
+	reset_libcw ();
+	atexit (close_libcw);
+
 	cw_register_keying_callback (keyingevent, NULL);
 
 	debug ("Device used: %s", cwdev->desc);
